RecordFunction in Dispatcher

ilia-cher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#37587 RecordFunction in Dispatcher**

Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

Commit:
[wip] RecordFunction in Dispatcher

Summary:
Lifting RecordFunction up into the dispatcher code

[ghstack-poisoned]

Commit:
Update on "[wip] RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

[ghstack-poisoned]

Commit:
Update on "[wip] RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

[ghstack-poisoned]

Commit:
Update on "[wip] RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

[ghstack-poisoned]

jamesr66a commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -283,11 +287,85 @@ namespace detail {
   const KernelFunction& kernel = dispatch_(dispatchTable, dispatchKey);
+
+  // RECORD_FUNCTION macro will attempt to box the arguments only if we have active observers enabled
+  RECORD_FUNCTION(op.schema().name(), std::vector<c10::IValue>{get_ivalue_copy(args)...}, at::FunctionSequenceNumber::peek());
```
Isn't this going to `RECORD_FUNCTION` for every key in the dispatch chain? Is that desired?

ilia-cher replied:
yes, that's the plan

jamesr66a commented on tools/autograd/gen_variable_type.py:
```diff
@@ -344,7 +340,6 @@
 TORCH_INTERNAL_ASSERT(op);
-RECORD_FUNCTION("${name}", std::vector<c10::IValue>({${input_names}}), Node::peek_at_next_sequence_nr());
 return c10::Dispatcher::singleton().callUnboxedRedispatch<${ret_and_arg_types}>(${profiled_dispatch_args});
 """)
```
Wrappers should probably be deleted if we're entirely lifting `RECORD_FUNCTION` into the dispatcher

ilia-cher replied:
yes, I will remove them later, we have some code that uses them (but they become functionally no-op)

Commit:
Update on "[wip] RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

[ghstack-poisoned]

ilia-cher commented:
(not ready for review yet)

Commit:
Update on "[wip] RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

[ghstack-poisoned]

Commit:
Update on "[wip] RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

[ghstack-poisoned]

Commit:
Update on "[wip] RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

[ghstack-poisoned]

smessmer commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -281,13 +283,54 @@ class CAFFE2_API OperatorHandle final {
+  // them into the callbacks and also into the kernel call
+  at::RecordFunction guard(at::RecordScope::FUNCTION);
+  if (guard.active_) {
+    // Stack-based record function with scope lifetime can be markee as 'current'
```
nit: "marked"

smessmer commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -281,13 +283,54 @@ class CAFFE2_API OperatorHandle final {
 namespace detail {
 template<class... Args> inline void unused_arg_(const Args&...) {}
+
+template <typename T, std::enable_if_t<c10::impl::not_ok_to_box<T>::value>* = nullptr>
```
oh so you're skipping arguments here? Can the observers handle that?

ilia-cher replied:
for observers we don't need to convert all inputs, only the ones that can be converted (typically tensors, scalars)

smessmer commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -281,13 +283,54 @@ class CAFFE2_API OperatorHandle final {
+  push_ivalue_copy(stack, item);
+}
+template<typename Item, typename... Rest>
+void convert_to_ivalue_vector(std::vector<c10::IValue>& stack, const Item& item, Rest... other_items) {
```
Can we move this logic to the boxing.h file, maybe call it `boxArgumentsIntoStack()` and then also call it from `boxAndCallBoxedFunc` to reuse the code?

dzhulgakov replied:
you probably want to rename `push_ivalue_copy` and `convert_to_ivalue_vector` into something more scary looking as it can randomly skip arguments. It's ok for profiling but pretty scary for other stuff

smessmer commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -281,13 +283,54 @@ class CAFFE2_API OperatorHandle final {
+      std::vector<c10::IValue> stack;
+      detail::convert_to_ivalue_vector(stack, args...);
+
+      guard._before(op.schema().name(), stack, at::FunctionSequenceNumber::peek());
```
Passing in `op` might be a bit better here, observers can get the name if that's what they want, or they can get the full schema if they need more info. Having the `op` handle, they could do all kinds of stuff and are more powerful.

ilia-cher replied:
let me see if I could it do it here in this PR, but for now, having a name as a string should be ok, we can extend the API further if we need

smessmer commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -316,6 +361,10 @@ inline void Dispatcher::callBoxed(const OperatorHandle& op, Stack* stack) const
   const KernelFunction& kernel = dispatch_(dispatchTable, dispatchKey);
+
+  // using already existing stack to record function execution in observers
+  RECORD_FUNCTION(op.schema().name(), *stack, at::FunctionSequenceNumber::peek());
```
if we do it manually in `callUnboxed`, I'd rather do it manually here doo. It's weird that one location uses macros and the other one doesn't. You can probably factor out common code between the record_function implementation in `callUnboxed` and here into subfunctions.

ilia-cher replied:
makes sense

smessmer reviewed (changes requested):
I think this is the right approach. Some comments with requests for changes.

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

ezyang commented:
This looks like it's still failing builds. Do you want reviews now, or later?

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

ilia-cher commented:
> This looks like it's still failing builds. Do you want reviews now, or later?

I need to address Sebastian's comments, will update, please do not review at the moment

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

dzhulgakov commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -281,13 +283,54 @@ class CAFFE2_API OperatorHandle final {
 template<class... Args> inline void unused_arg_(const Args&...) {}
+
+template <typename T, std::enable_if_t<c10::impl::not_ok_to_box<T>::value>* = nullptr>
+inline void push_ivalue_copy(std::vector<c10::IValue>& stack, const T& v) {}
```
it's kind of shitty that we just skip the arg and it makes looking at results weird. How about you push a string literal `'<cannot box it>'`

ilia-cher replied:
in this case every observer needs to explicitly know about this special value and filter it out;
I will rename the function to make it more obvious that it doesn't convert all arguments

ilia-cher replied:
to be more clear, arguments that are not boxable are not only unsupported in observers but also in some sense less interesting since observers wouldn't be able to do much with them anyway even if we could pass them, so I start from the assumption that observers want to receive only the ivalues they know how to handle

dzhulgakov replied:
Hm, still I think it's better to pass the same number of positional arguments as in schema to observer than to drop some of them. I literally mean passing `<can't box>` as a string literal, imho it's a better failure mode

dzhulgakov commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -281,13 +283,54 @@ class CAFFE2_API OperatorHandle final {
+    // Stack-based record function with scope lifetime can be markee as 'current'
+    // thread local record function; used to track nested recorded scopes
+    guard._setCurrent();
+    if (guard.needs_inputs_) {
```
put UNLIKELY or whatever is appropriate

ilia-cher replied:
we usually do need inputs though, both for profiler and fleet-wide observers

dzhulgakov commented on tools/autograd/gen_variable_type.py:
```diff
@@ -346,7 +342,6 @@
 PROFILE_DISPATCH_UNBOXED = CodeTemplate("""\
```
add at least a TODO to clean up them?

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

dzhulgakov commented on aten/src/ATen/FunctionSequenceNumber.h:
```diff
@@ -0,0 +1,19 @@
+// A simple thread local enumeration, used to link forward and backward pass
+// ops and is used by autograd and observers framework
+
+class TORCH_API FunctionSequenceNumber {
```
nit: it can be just a namespace

dzhulgakov commented on aten/src/ATen/ThreadLocalState.cpp:
```diff
@@ -29,6 +33,8 @@ void ThreadLocalState::setThreadLocalState(
 
   _setTLSCallbacks(state.callbacks_);
 
+  at::enableRecordFunction(state.observers_enabled_);
```
just curious - why do you put it separate and don't use threadlocaldebuginfo and those setters/getters you had

ilia-cher replied:
ThreadLocalDebugInfo is for range (scope) based settings and this one is per thread, and having it close to the definition of record function seemed more reasonable

dzhulgakov commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -287,13 +289,53 @@ class CAFFE2_API OperatorHandle final {
+
+      // if we could convert all the arguments, also pass the stack into the kernel call
+      if (stack.size() == sizeof...(args)) {
+        return kernel.template callUnboxedWithStack<Return, Args...>(op, stack, std::forward<Args>(args)...);
```
sorry, I don't get why you need this call with a different codepath if the guard is active? why not just fall through and user regular call below?

ilia-cher replied:
that's something that we discussed earlier with @smessmer - this is to be able to reuse the stack that we created for the observer in case we call boxAndCallBoxedFunc later

ilia-cher commented:
@dzhulgakov just finished with the pre-requisite PR, this PR is not ready yet, I'll update it and will ping you and other reviewers; also thanks for reviewing!

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

ilia-cher commented:
@smessmer @dzhulgakov @ezyang - PR is ready for another look

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

smessmer commented on aten/src/ATen/core/boxing/impl/boxing.h:
```diff
@@ -16,25 +16,79 @@ namespace impl {
+  return boxArgumentsIntoStack(stack, other_items...) && res;
+}
+
+template<class Result>
```
nit: those three variants could be combined into one with a `if_constexpr`, which would make it a bit easier to read in my opinion.

smessmer commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -294,6 +298,28 @@ inline Return Dispatcher::callWithDispatchKey(const OperatorHandle& op, Dispatch
+  at::RecordFunction guard(at::RecordScope::FUNCTION);
+  if (guard.active) {
+    if (guard.needs_inputs) {
+      std::vector<c10::IValue> stack;
```
if this is a perf problem, we could think about using a thread_local std::vector and reuse it instead of allocating one on each call. But not sure if perf is important here, maybe not because we're inside the guard.

ilia-cher replied:
"active" is generally false most of the time

smessmer replied:
ok. Can we add C10_UNLIKELY to the if condition?

smessmer reviewed (approved):
This looks good to me. Please coordinate landing this with @bhosmer, your PRs are going to conflict. If you decide to land yours first, @bhosmer will be able to clean up much of the complexity this PR is introducing. If you decide to land @bhosmer's first, you could clean it up yourself before landing. Up to you guys.

dzhulgakov reviewed (commented):
Looks good, but I'm still curious about that comment

ilia-cher commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -294,6 +298,28 @@ inline Return Dispatcher::callWithDispatchKey(const OperatorHandle& op, Dispatch
+
+      // if we could convert all the arguments, also pass the stack into the kernel call
+      if (boxed_all_args) {
+        return kernel.template callBoxedOrUnboxed<Return, Args...>(op, stack, std::forward<Args>(args)...);
```
it's a trade off between code simplicity and a small win of not boxing twice; I'm mostly neutral about it here, but @smessmer might have something more to add and maybe give an answer regarding if callBoxedOrUnboxed going to be used later elsewhere

smessmer replied:
Yeah this only exists to prevent having to box arguments twice. I think doing an extra boxing adds a small perf overhead that probably doesn't matter much and adding this `callBoxedOrUnboxed` helper adds a small complexity overhead that probably doesn't matter much. Overall, I prefer avoiding the perf overhead, but it's not a strong opinion.

bhosmer replied:
@dzhulgakov the main motivation for this is just that in general, we want to minimize boxed/unboxed transitions as we move down the redispatch chain. This situation here is novel because the profiling logic is introducing a unilateral boxing step - i.e. one that doesn't arise strictly from how an op's kernels were implemented - so it's the first time we're in the position of having both forms available. Like @smessmer I don't think it's critical, but my reasons for preferring it this way are basically
* It's not impossible that we might later want to do a similar thing in other contexts (i.e., internal dispatcher logic provokes a boxing - or more significantly, unboxing - step that we want to reuse, or possibly that we add it as part of the redispatching API)
* I think we should treat the dispatcher as a big bundle of hot paths and err on the side of conserving work when it can be done without making the logic too ugly.  I think this clears that bar, but wdyt?

dzhulgakov replied:
I don't have strong opinion either. I just don't feel shaving off extra boxing from profiling codepath is that critical perf wise. However, since it's already implemented, we can keep it. As you said, it's good to have all tools in the toolbox

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

ilia-cher commented on aten/src/ATen/core/boxing/impl/boxing.h:
```diff
@@ -16,25 +16,79 @@ namespace impl {
+
+template <typename T, std::enable_if_t<!c10::impl::ok_to_box<T>::value>* = nullptr>
+inline bool pushIValueOrCannotBox(std::vector<c10::IValue>& stack, const T& v) {
+  torch::jit::push(stack, "cannot box");
```
chatted with @bhosmer earlier

ezyang replied:
It might be better not to push a string here, IIUC this will turning into an owning string in the IValue which involves a dynamic allocation (which really isn't necessary here)

ezyang replied:
Also, if I understand correctly, previously, if we incorrectly annotated a function as supporting full c10 dispatch, we would have gotten a compile time error saying that boxAndCallBoxedFunc did not support it. With the changes here, now we no longer get a compile time error in this case. This seems a bit dangerous, as if you use this codepath to attempt to actually call a boxed function (rather than RecordFunction), you will get garbage inputs.

ilia-cher replied:
sg, keeping the function here for now, and renaming to boxArgumentsOrCannotBoxIntoStack

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -294,7 +298,34 @@ inline Return Dispatcher::callWithDispatchKey(const OperatorHandle& op, Dispatch
   const KernelFunction& kernel = dispatch_(dispatchTable, dispatchKey);
-  return kernel.template call<Return, Args...>(op, std::forward<Args>(args)...);
+
+  if (dispatchKey != DispatchKey::BackendSelect) {
```
This special case looks really suspicious. Why do you have to special case BackendSelect?

ilia-cher replied:
I see in stack traces that for ops we call callWithDispatchKey first with BackendSelect
and then callWithDispatchKey with an actual dispatchKey

ezyang replied:
Yes, this is what BackendSelect does: when you do a factory function, you hit BackendSelect first, and then BackendSelect figures out what the correct Backend is and then redispatches you there.

It seems clear you don't want to profile twice in this case. But there should be a very simple fix for this: make *BackendSelect* redispatch bypass record function. Then you can avoid this special case entirely.

ilia-cher replied:
what do you mean by "make BackendSelect redispatch bypass record function"? Right now we don't use RecordFunction in `redispatch`

ezyang replied:
Sorry, I wasn't using redispatch in the technical sense here. Here's what BackendSelect generates right now:

```
Tensor _cudnn_init_dropout_state(double dropout, bool train, int64_t dropout_seed, const TensorOptions & options) {
  static auto op = c10::Dispatcher::singleton()
    .findSchemaOrThrow("aten::_cudnn_init_dropout_state", "")
    .typed<Tensor (double, bool, int64_t, const TensorOptions &)>();
  DispatchKey _dk = options.computeDispatchKey();
  return op.callWithDispatchKey(_dk, dropout, train, dropout_seed, options);
}
```

You should replace callWithDispatchKey with a call that bypasses RecordFunction.

ilia-cher replied:
I assume, you mean passing a flag into callWithDispatchKey? e.g. extra boolean whether to use record function

ezyang replied:
Mm, the exact API is up to you. It might be better to make it an optional template parameter so that it is obvious that it is statically known (and so the condition will be optimized away).

bhosmer replied:
@ezyang it feels kind of inverted to make avoiding double counting the responsibility of kernels, doesn't it? 

If we can say something simple like: we only want to record invocations that run "real" kernels (dispatch key = CPU, GPU, ...), and not others, then wouldn't it be better to have some predicate `shouldRecord(dk)` on dispatch keys that we can call from the e.g. site of @ilia-cher 's original if test? Centralizing the knowledge in the dispatcher would
* avoid the entanglement where "nonterminal" kernels (e.g. BackendSelect, but maybe others as well - now or later) need to know about RecordFunction and double counting
* avoid muddy reentrancy situations e.g. where suppressing RecordFunction in BackendSelect winds up undercounting reentrant invocations issued from whatever kernel is selected. (Because in the current solution, we're counting the BackendSelect invocation, not the subsequent kernel invocation.)

ezyang replied:
There are a few reasons I don't like it:

1. Now we have to do an extra computation and test to see if we do record function. If kernels decide whether or not to suppress record function or not, you can eliminate the conditional. This is my most important reason.
2. Testing on dispatch key is incompatible with fallthrough optimization. The semantics of fallthrough are that it should be equivalent to having invoked on the dispatch key, and then the kernel redispatched to the next kernel. With the fallthrough optimization, there is never actually an invocation of callWithDispatchKey with the dispatch key that was supposed to fall through, which means if both dispatch keys were intended to RecordFunction, then you will lose one of the observer calls. If you say that dispatcher calls can decide if they record function or not, then you can simply define the reference implementation of fallthrough to bypass profiling.

I also feel that RecordFunction suppression morally belongs at these call site. A call site that is redispatching knows if it did "a lot of" work or not very much work. When a call site has done very little work, it should skip RecordFunction and let the redispatch attribute it costs to the enclosing cost center. If initial calls into the dispatcher always do RecordFunction, there will always be a reasonable cost center to attribute the performance too.

ilia-cher replied:
re. 1 - isn't it a trivial check (uint8_t == check?)

ilia-cher replied:
a qq. - BackendSelect is not supposed to happen in callBoxed?

bhosmer replied:
@ezyang @ilia-cher sorry for the late reply -
1. true, it's an extra conditional, but only when RecordFunction is on. (I just realized @ilia-cher 's code here doesn't do it this way, but I think it could - i.e., line 336 here could be moved to after line 344.) So we'd be making profiling incrementally faster by avoiding the conditional, not all dispatch.
2. hmm but isn't it incompatible with fallthrough to host profiling in callWithDispatchKey, irrespective of filtering? You'll miss the obvservations regardless. If I follow your description, I think you're just saying that if the convention is for kernels to decide, then you avoid this by... having kernels instead of fallthrough? I'm probably missing your point here.

Re morality, I agree that call-site-decides is one way to bucket a multi-level dispatch chain together. I think for things like BackendSelect, where as a class of kernels they're going to be predictably lightweight, you have the information you need to make a centralized decision - but maybe it's more fraught to assume the same for keys with more diverse kernels.

Anyway my most important reason :) is wanting to keep arcane dispatcher knowledge from leaking out into kernels. Like, will each new set of kernels for utility dispatch keys (e.g., Meta) need to remember to turn profiling off? Will observers ever be used for new things for which would kernels would need to make new decisions? But I guess we can always go back and revisit if any of that starts to become problematic.

ilia-cher replied:
@bhosmer do you mean that it's better in your opinion to always create `at::RecordFunction guard(at::RecordScope::FUNCTION);`, then check `if (C10_UNLIKELY(guard.active))` and then for `useRecordFunction && op.operatorIterator_->op.isObserved()` ? wouldn't it be better to just avoid this path completely if useRecordFunction is turned off or the op is not observable at all?

ilia-cher replied:
also I'm more inclined to agree with @bhosmer as it does seem more to expose the detail of dispatcher to the interface,  will wait for @ezyang to address

bhosmer replied:
> @bhosmer do you mean that it's better in your opinion to always create `at::RecordFunction guard(at::RecordScope::FUNCTION);`, then check `if (C10_UNLIKELY(guard.active))` and then for `useRecordFunction && op.operatorIterator_->op.isObserved()` ? wouldn't it be better to just avoid this path completely if useRecordFunction is turned off or the op is not observable at all?

Well, in my mind useRecordFunction is the thing we're trying to avoid, right? (Aside - if we do go with useRecordFunction, I'd make it a template param per @ezyang's suggestion. The passed arg is always constant AFAICT, and the specialization should give us a nice perf bump.)

Anyway though, what I was noting was that the logic went from
```
at::RecordFunction guard(at::RecordScope::FUNCTION);
if (C10_UNLIKELY(guard.active)) {
  do profiling stuff 
}
```
to
```
if (dispatchKey != BackendSelect) { 
  at::RecordFunction guard(at::RecordScope::FUNCTION);
  if (C10_UNLIKELY(guard.active)) {
    do profiling stuff 
  }
}
```
which adds the comparison unconditionally to a very hot hot path, which is definitely something we'd want to avoid if we can. But if instead we move the new test inside the record function block (logically equivalent I think), then we're  only adding overhead when we're profiling:
```
at::RecordFunction guard(at::RecordScope::FUNCTION);
if (C10_UNLIKELY(guard.active)) {
  if (dispatchKey != BackendSelect) { // or e.g. shouldRecord(dispatchKey) or whatever
    do profiling stuff 
  }
}
```

ezyang commented on aten/src/ATen/core/boxing/KernelFunction.h:
```diff
@@ -86,6 +86,14 @@ class CAFFE2_API KernelFunction final {
 
+
+  /**
+   * Like call, but with a stack of inputs already passed
```
This needs a more detailed description of the calling convention. Does the `stack` get mutated to have the updated return arguments unconditionally, or does it depend on if you took the boxed or unboxed call path?

ilia-cher replied:
I'll expand the comment; it's for the boxed arguments and will be used if the boxed path is chosen

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

ezyang commented:
This looks good! I stand by my suggestion to remove the BackendSelect special case, but everything else looks reasonable.

ilia-cher commented:
I noticed the issue with the current implementation - it boxes everything and that slows it down compared to the current approach when we only pass tensors and scalars (when we codegen RECORD_FUNCTION macro).
Ideally we'd want to pass all the arguments and the boxing seems to be pretty unavoidable (we register observers in runtime and can't pass the parameter pack into them), so we could either accept the slow down of the RecordFunction path (it is normally heavily sampled, unless we do explicit profiling) or avoid boxing everything - e.g. expanding the list to tensors, scalars, primitive types (int, bool, float, possibly arrays of them) would be useful for observers, and won't require boxing everything or passing "cannot box" string

ezyang commented:
In this case, if you are *sure* you don't need non-Tensor arguments, I would probably just ditch the "reuse the boxed arguments" optimization and just only box tensors/tensorlists. But it seems like if you're trying to capture operator dimensions as part of observers, the other arguments on the function are pretty important too for understanding how things are being used, as those arguments can greatly impact what algorithm is actually being exercised under the hood...

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

ilia-cher commented:
Rebase
We'll keep the existing boxing logic (box all arguments) as we indeed need not only tensors in the observers

ilia-cher commented:
(looks like changes in tracer introduced some build errors, looking into them)

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

ilia-cher commented:
addressing comments, adding filtering of observed ops

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

bhosmer commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -350,15 +382,30 @@ inline Return Dispatcher::redispatch(const TypedOperatorHandle<Return (Args...)>
   const auto& kernel = entry.lookup(dispatchKey);
-  kernel.callBoxed(op, stack);
+
+  if (entry.isObserved()) {
```
Note: the absence of `useRecordFunction` here introduces an API inconsistency between `callBoxed` and `callWithDispatchKey` - one lets you control RecordFunction and the other doesn't. If we go with the `useRecordFunction` approach, I'm thinking we should add it to `callBoxed` as well (we can avoid callsite churn by defaulting it). 

But also, if you're inclined to dislike call-site-decides because of leakage, here's some more :P

ilia-cher replied:
I was wondering in what cases callBoxed would be called with BackendSelect, or is it for the symmetry

bhosmer replied:
Yeah, just for the symmetry. Unless there's some history that I'm missing, there's no reason to think we don't want the same knobs on boxed calls as unboxed. 

Note: I wouldn't necessarily burden your PR with the addition - I just wanted to note it in the context of the discussion about filtering BackendSelect et al.

bhosmer commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -325,10 +329,38 @@ template<class... Args> inline void unused_arg_(const Args&...) {}
+    // Note: for perf reasons we wouldn't want to pass arguments into
+    // the function call or prematurely box them
+    at::RecordFunction guard(at::RecordScope::FUNCTION);
+    if (C10_UNLIKELY(guard.active)) {
```
I think you can put the dispatch key filter test inside this `if`, and also move the `op.isObserved()` test down here.

bhosmer reviewed (commented):
Hi Ilia - so after discussing with @ezyang, we came to two conclusions:
* at some point in the future, we might want to move to a "true" callee-profiling setup, which wouldn't be in the dispatcher at all - instead, we'd use metaprogramming behind the registration API to instrument kernels as they were registered. This would have a few benefits besides incrementally better precision (full per-kernel granularity with no dispatch overhead; direct calls wouldn't be missed), but it's not compelling enough to disrupt the current approach and sink a bunch of work into now.
* for now, the best approximation to this future state is to revert the `useRecordFunction` param and go back to (a variation of) your original approach, where `callWithDispatchKey` does a dynamic test on dispatch key. Two changes seem worthwhile though:
  * the new test, along with the test on `op.isObserved()`, can be done inside the `if (C10_UNLIKELY(guard.active)) { ... }` block, as noted inline. This will at least keep the testing overhead to a minimum.
  * instead of doing a direct test on `BackendSelect`, let's wrap it in a (inline) filter function that takes a dispatch key and returns a bool, say `shouldRecord()` or something. That'll give us a place to put additional filtered keys as they come up.

Otherwise LGTM (I happened to notice a nit from @smessmer as I scrolled past, don't know if that's a remaining TODO...) - just re-request or gimme a ping once these changes are up and we should be good to go!

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

ilia-cher commented:
(rebase)

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

ilia-cher commented:
@bhosmer ready for another look, I was a bit unsure about @smessmer nit comment as it would involve using of lambdas which seemed unnecessary, but if you think it's necessary I'll switch to using if_constexpr

bhosmer commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -350,14 +387,28 @@ inline Return Dispatcher::redispatch(const TypedOperatorHandle<Return (Args...)>
       args...
     );
-  return callWithDispatchKey<Return, Args...>(op, dispatchKey, args...);
+  // do not use RecordFunction on redispatch
```
Hmm is there a reason to break symmetry here? I think we want `redispatch` to be exactly equivalent to `callWithDispatchKey` on the computed key - basically just a convenience method. Introducing this logic change seems like a potential source of hard-to-track-down inconsistencies in recording results. 

If the change is to avoid double-counting some key other than BackendSelect, maybe that key should be added to `shouldRecord`? cc @smessmer maybe you guys have talked about this already...

ilia-cher replied:
the reason I do it this way here is that I don't want to run RecordFunction and we don't want to pass a parameter into callWithDispatchKey (as we discussed). Since it's two lines of code I just called the kernel itself

bhosmer replied:
Yeah, I definitely have no problem with the implementation. My reaction was to the semantics - my default assumption was that we want `redispatch(op, args)` to have the exactly same behavior as the equivalent `callWithDispatchKey(op, key, args)` call - w.r.t. RecordFunction and everything else. 

But the other POV makes sense too - that a natural consequence of the meaning of "redispatch" is: no (re)recording, so they really are different. 

It's not totally clear to me that the suppression shouldn't move into `shouldRecord`, but it's not a discussion we need to have before you land this 😄 thanks for being patient!

bhosmer reviewed (approved):
Looks good! one ~trivial naming idea and one nontrivial question that might be a suggested change (the redispatch thing). Other changes look great, thanks!

ilia-cher commented:
will move and rename boxArgumentsIntoStack

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]

Commit:
Update on "RecordFunction in Dispatcher"


Summary:
Lifting RecordFunction up into the dispatcher code

Differential Revision: [D21374246](https://our.internmc.facebook.com/intern/diff/D21374246)

[ghstack-poisoned]
